Solutions/Corvus.Identity.Azure/Corvus/Identity/ClientAuthentication/Azure/Internal/IKeyVaultSecretCache.cs

```diff
@@ -0,0 +1,61 @@
+﻿// <copyright file="IKeyVaultSecretCache.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Corvus.Identity.ClientAuthentication.Azure.Internal
+{
+    using System.Diagnostics.CodeAnalysis;
+
+    /// <summary>
+    /// Caches secrets to avoid repeated lookups in Azure Key Vault.
+    /// </summary>
+    internal interface IKeyVaultSecretCache
+    {
+        /// <summary>
+        /// Retrieve a cached secret it one is available for the specified combination of vault
+        /// name, secret name, and client identity.
+        /// </summary>
+        /// <param name="vaultName">The name of the vault containing the secret.</param>
+        /// <param name="secretName">The name of the secret.</param>
+        /// <param name="clientIdentity">
+        /// The identity that will be used to access the key vault if the secret is not cached.
+        /// </param>
+        /// <param name="secret">The secret, or null if the secret is not available.</param>
+        /// <returns>
+        /// True if the cache contains this secret.
+        /// </returns>
+        bool TryGetSecret(
+            string vaultName,
+            string secretName,
+            ClientIdentityConfiguration? clientIdentity,
+            [NotNullWhen(true)] out string? secret);
+
+        /// <summary>
+        /// Adds a secret to the cache.
+        /// </summary>
+        /// <param name="vaultName">The name of the vault containing the secret.</param>
+        /// <param name="secretName">The name of the secret.</param>
+        /// <param name="clientIdentity">
+        /// The identity that was used to access the key vault.
+        /// </param>
+        /// <param name="secret">The secret.</param>
+        void AddSecret(
+            string vaultName,
+            string secretName,
+            ClientIdentityConfiguration? clientIdentity,
```

Comment:
Would it be reasonable to default these nullable parameters to null (reordering if necessary)? If the thing is nullable for optionality?

Reply:
I have added comments explaining why it might be null. This value comes from `KeyVaultSecretConfiguration.VaultClientIdentity`, and if you don't set the `VaultClientIdentity` that signifies that you want the service identity (whatever the application may have configured that to be; might be an Azure Managed Identity, but when debugging locally it's more likely to be either the default credential (which falls back to `az`) or an Azure AD app with client secret).

I don't want to give this a `default` because this is an `internal` interface designed to be called from exactly one place in the code, and it would be an error for that call site to fail to pass the `VaultClientIdentity` from the relevant `KeyVaultSecretConfiguration`. The non-defaultability of this correctly captures the code invoking this method should always positively pass something, and sometimes that something will be `null`.